pkg/controller/common.go

```diff
@@ -247,7 +247,7 @@ func newLockableEntry(entry interface{}) *lockableEntry {
 }
 
 func shouldRequeueReplicaOperation(isReplicaGarbageCollection bool, err error) bool {
-	return !isReplicaGarbageCollection || !errors.Is(err, context.Canceled)
+	return !isReplicaGarbageCollection && err != nil && !errors.Is(err, context.Canceled)
```

Comment:
`err == nil` check would be redundant to `shouldRequeueReplicaOperation` if `shouldRequeueReplicaOperation` remains as `!isReplicaOperation || context.Is(err, context.Canceled)` because if `err == nil`, `context.Is(err, context.Canceled)` would evaluate to `false`.

Reply:
For the case `err == nil` and `!isReplicaGarbageCollection`, `shouldRequeueReplicaOperation` will return `true`. Maybe change the order to `context.Is(err, context.Canceled) || !isReplicaGarbageCollection`?

Reply:
Ah, I see what you mean; I forgot about the `!isReplicaOperation` evaluation; my bad. Changing the order would not help as it is chained with an OR. So, we would need `err == nil` check as you proposed earlier before closing the workflow.

Reply:
Regardless of whether there is an error or not, the operation should not be requeued if there is a `isReplicaGarbageCollection`. Thus, the proposed check with OR will not work. I don't see what the issue with the existing check is.

Reply:
> This entirely changes the logic of determining whether a replica operation should be requeued or not. The intent of this check was to ensure that all replica operation can be requeued on all errors unless the operation is replica garbage collection that had been aborted due to context cancellation (i.e. `isRepliaGarbageCollection && errors.Is(err, context.Canceled)`).
> 
> That is. we should requeue operation when `!(isReplicaGarbageCollection && errors.Is(err, context.Canceled))`, which is identical to `!isReplicaGarbageCollection || !errors.Is(error, context.Canceled)`).
> 
> Speaking of which, there seems to be a bug in `triggerReplicaGarbageCollection`, as it creates `workflowCtx` with `context.Background()`, breaking the logic for this check.
> 
> https://github.com/kubernetes-sigs/azuredisk-csi-driver/blob/3f5114357ac4c342eef8fd174d1e0c706edb11d3/pkg/controller/replica.go#L156
> 
> 
> It should create `workflowCtx` with `deletionCtx` instead in order for us to be able to assess `shouldRequeueReplicaOperation` properly.
> https://github.com/kubernetes-sigs/azuredisk-csi-driver/blob/3f5114357ac4c342eef8fd174d1e0c706edb11d3/pkg/controller/replica.go#L150

This is not accurate. We should requeue when `!(isReplicaGarbageCollection || errors.Is(err, context.Canceled))` which is  `!isReplicaGarbageCollection && !errors.Is(error, context.Canceled)`).

Reply:
I think it may be `replicaOperation.ctx` that may be bringing the confusion here.
https://github.com/kubernetes-sigs/azuredisk-csi-driver/blob/3f5114357ac4c342eef8fd174d1e0c706edb11d3/pkg/controller/common.go#L147-L152

This context may be seen as a way to control the operations in the operation queue, (which we may as well add support for in the future but this would belong to a separate PR after evaluating further), but this was only added as a way to support workflow logging in operation queue.

That is, we should be requeueing all failed replica operations except for the garbage collection that was cancelled.

Reply:
> Oh, you are saying that context.Canceled is only relevant for garbage collection? I didn't realize the cleanup is happening through the same queue.

Yup, for now, we only utilize context errors for garbage collection operations. We may expand usage of context used in the operation queue but this would belong to a separate PR.

Reply:
So, if we only use it in garbage collection, do we need to keep track of `isReplicaGarbageCollection` at all?

Reply:
I change the logic to `err != nil && !(isReplicaGarbageCollection && errors.Is(err, context.Canceled))` and add comment, so we don't need to check whether the err == nil or not before `shouldRequeueReplicaOperation` and I feel it's more readable. Hope this makes sense to everyone as well.

Reply:
Makes sense to me 👍